### PR TITLE
EKF: Fix magnetometer fusion and covariance handling

### PR DESCRIPTION
**Pull request by priseborough**
Poor heading stability when stationary for longer time periods had been observed. During the investigation a number of deficiencies were identified:

1) A bug in the original derivation that meant a tan function was being used instead of an atan function in the derivation of the observation Jacobians and Kalman gains for the direct heading and declination fusion.
2) The covariance matrix terms for inactive states were not being explicitly controlled
3) The various fusion operations did not have explicit protections against modifiying inactive states
4) The covariance terms for inactive states were not being explicitly set when changing modes
5) The method of preventing uncontrolled covariance growth on magnetic field states relied on hard variance constraints and did not control the covariance terms. This could result in numerical instability

These deficiencies have been rectified, and placeholder to enable management of wind state covariance terms has been added to reduce the likelihood of similar errors being introduced when airspeed and sideslip fusion is added.

Ground testing has been completed and flight testing will be completed shortly  for operation using the three available magnetometer fusion mode parameter settings. The following figure shows the declination for the estimated NE field after 10 minutes of static operation with noisy GPS. The declination drift is acceptable, so the declination fusion during aiding has been turned off by default (it can be turned on by setting EKF2_DECL_TYPE bit position 2 to true).

without declination fusion

The following figures are taken from a flight with EKF2_MAG_TYPE = 1 which turns off use of the magnetic field states and uses a simple magnetic heading observation model. and velocity and position innovations were zero mean throughout with no unexpected vehicle movement observed during the flight.

GPS velocity and position innovations remained within limits and were centred about zero other than towards the end of flight when it was flown into an area with significant GPS masking and the GPS solution was glitching

heading fusion - vel innov

heading fusion - pos innov

yaw gyro bias and scale factor variation was normal

heading fusion - dangz bias

heading fusion - dangz scale

The following figures were taken with EKF2_MAG_TYPE = 2 which forces use of the 3-axis magnetometer fusion and EKF2_DECL_TYPE = 3 which means that fusion of the declination was not performed after GPS fusion started. This was done to test the stability of the earth field estimation. The flight was performed in a continuous hover with occasional yaw orientation changes to try and maximise declination drift.

The estimated declination during the flight was relatively constant with abut 5deg of change at the end of flight when the yaw orientation was changed.

mag3d fusion - declination

Velocity, position and mag innovations were normal

mag3d fusion - vel innov

mag3d fusion - pos innov

mag3d fusion - mag innov

IMU Z gyro bias and scale factor variation was normal

mag3d fusion - dangz bias

mag3d fusion - dangz scale

### CODE REVIEW

**Comment by priseborough** [comment]
Update - I cannot get it to log with EKF2_MAG_TYPE = 1 as the logging program falls over. I have also had a crash when switching flight modes when the pixracer locked up so am concerned about a run-time error.

Additionally the numerical errors with the covariance calculation when hovering for long periods has not been eliminated, so further investigation is required.

**Comment by priseborough** [comment]
Hee is the nsh terminal output just prior to , and just post arming. The sdlog2 program falls over

nsh> [Ksdlog2 status
sdlog2: extended logging: OFF
sdlog2: time: gps: 1455596422 seconds
sdlog2: not logging
nsh> [K
nsh> [Ktelem> TAKEOFF DETECTED
teAssertion failed at file:armv7-m/up_hardfault.c line: 184 task: sdlog2
sp:     20003080
IRQ stack:
  base: 200030cc
  size: 000002e8
20003080: 080f11f4 000000b8 080ae799 00000010 20002dcc 00000003 00000000 080b2f01
200030a0: 080b2eed 080bf3c7 e000ed24 080bc1a5 00000000 447a0000 00000810 20001eb8
200030c0: 08088b71 080b2e71 200181dc 00000000 00000000 00000000 00000000 00000000
sp:     200182b0
User stack:
  base: 20018ad0
  size: 00000ce4
200182a0: 00000000 00000000 00000000 080b1f55 00000000 100079c0 00000800 080b2e1f
200182c0: 20017770 100079c0 10007b50 080aeabf 20001eac 20001eb8 08088b71 00000000
200182e0: 00000001 200185e0 20001e7c 200185e0 00000007 080890ff 00000000 0000003c
20018300: 00000000 00000000 00000000 00000000 00000000 200183e8 00000000 20017ac8
20018320: 00000001 00000000 2001840f 080eb544 200184b4 00000000 00000000 ffffffe9
20018340: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 20018430
20018360: 00000000 20018438 00000000 20000744 0000000a 20018448 00000000 00000000
20018380: 2000077c 00000000 2000077c 00000001 1000416d 00000000 00000000 ffffffe9
200183a0: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
200183c0: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
200183e0: 00000000 2000077c 00000000 00000000 0000000b 080af009 080af31e 61000000
20018400: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
20018420: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
20018440: 00000000 080af319 20000744 1000416d 0000000d 080b8951 080b8859 200177d0
20018460: 00000000 00000000 20017acc 080af319 20017ac8 1000416d 1000416d 080b99a7
20018480: 20017ac8 080b066b 20009540 0000000a 20017ac8 200185e0 20001e7c 080b0363
200184a0: 0000000a 0a017ac8 20001e1d 00000001 200185e0 0808928d 080ec980 0000002e
200184c0: 00000012 00000004 00000010 00000001 00000074 01000000 200186fc 00000000
200184e0: 00000000 00000001 00000001 08089813 2001851c ffffffff ffffffff ffffffff
20018500: ffffffff ffffffff ffffffff ffffffff 00002710 00000001 00000007 200096f0
20018520: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
20018540: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
20018560: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
20018580: 00000000 00000000 00000000 00000000 00000000 00000000 ffffffff 000095a3
200185a0: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
200185c0: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 ff000000
200185e0: 0717c640 00000000 e96e7564 563d8184 0000819d 00000000 0717c640 00000000
20018600: 3f70a3d8 3f3136b5 00000003 40973334 40cdd2f2 0000006b 00000007 00000000
20018620: 0717c640 00000000 3e5f3b65 3e5b22d2 3d3851ec bd89374c 3fcdf4e8 00000001
20018640: 0717c63f 00000000 751b7a0c 00052bdb 0000000a 00000000 00000001 00000000
20018660: 07195b24 00000000 02000c00 00000000 00000002 00000001 00000001 00000001
20018680: 00010000 00000001 00010000 00000101 40a03466 00000100 00000000 00000000
200186a0: 01000000 00000000 00000000 00010000 00000000 00000000 00000000 00000000
200186c0: 00000000 00000000 3eaa1fb0 00000000 00000000 00000000 00000000 00000000
200186e0: 00000000 00000000 00000000 01010000 00010101 00000000 00000000 00000001
20018700: ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff
20018720: ffffffff ffffffff ffffffff ffffffff ffffffff 00000004 ffffffff ffffffff
20018740: ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff
20018760: ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff
20018780: ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff 00000000 00000000
200187a0: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
200187c0: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
200187e0: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
20018800: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
20018820: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
20018840: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
20018860: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
20018880: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
200188a0: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
200188c0: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
200188e0: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
20018900: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
20018920: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
20018940: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
20018960: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
20018980: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
200189a0: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
200189c0: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
200189e0: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
20018a00: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
20018a20: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
20018a40: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
20018a60: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
20018a80: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
20018aa0: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
20018ac0: 00000000 080afc75 00000000 00000000 ffffffff a82464f2 00000060 80000cf0
R0: 20002ce0 20009540 20002c74 00000000 447a0000 00000810 20001eb8 08088b71
R8: 20001eac 20001e80 00000000 00000000 0000000b 200182b0 080b1f61 080b1f6e
xPSR: 81000000 BASEPRI: 00000000 CONTROL: 00000000
EXC_RETURN: ffffffe9

**Comment by priseborough** [comment]
The errors cannot be reproduced on other boards, so further testing of this PR is required

**Comment by tumbili** [comment]
@priseborough I'll review this!

File: EKF/mag_fusion.cpp [hunk 1]
```diff
@@ -651,50 +701,53 @@ void Ekf::fuseDeclination()
...
+
+	} else {
+		Kfusion[22] = 0.0f;
+		Kfusion[23] = 0.0f;
+	}
 
 	// calculate innovation and constrain
-	float innovation = atanf(t4) - _mag_declination;
+	float innovation = atanf(magE / magN) - _mag_declination;
 	innovation = math::constrain(innovation, -0.5f, 0.5f);
```
Comment (bugobliterator) [inline EKF/mag_fusion.cpp]: shouldn't this be -pi/2 to pi/2 constrain?

Comment (priseborough) [inline EKF/mag_fusion.cpp]: No.

magN will always be positive if it has come this far and the published declination will not exceed +-pi/2

**Pull request merged**